File: spring-boot-iot-system/src/main/java/com/ghlzm/iot/system/entity/Menu.java
package com.ghlzm.iot.system.entity;

import com.baomidou.mybatisplus.annotation.IdType;
import com.baomidou.mybatisplus.annotation.TableField;
import com.baomidou.mybatisplus.annotation.TableId;
import com.baomidou.mybatisplus.annotation.TableLogic;
import com.baomidou.mybatisplus.annotation.TableName;
import lombok.Data;

import java.io.Serial;
import java.io.Serializable;
import java.util.Date;

@Data
@TableName("sys_menu")
public class Menu implements Serializable {

      @Serial
      private static final long serialVersionUID = 1L;

      @TableId(value = "id", type = IdType.ASSIGN_ID)
      private Long id;

      private Long parentId;

      private String menuName;

      private String menuCode;

      private String path;

      private String component;

      private String icon;

      @TableField("meta_json")
      private String metaJson;

      private Integer sort;

      private Integer type;

      private Integer status;

      private Long createBy;

      private Date createTime;

      private Long updateBy;

      private Date updateTime;

      @TableLogic
      private Integer deleted;
}

File: spring-boot-iot-ui/src/api/menu.ts
import { request } from './request'
import type { ApiEnvelope, IdType, PageResult } from '../types/api'
import type { MenuTreeNode } from '../types/auth'

export interface Menu {
  id: IdType
  parentId?: IdType | null
  menuName: string
  menuCode?: string
  path?: string
  component?: string
  icon?: string
  metaJson?: string
  sort?: number
  type?: number
  status?: number
  createBy?: number
  createTime?: string
  updateBy?: number
  updateTime?: string
  children?: Menu[]
  hasChildren?: boolean
}

export interface MenuListParams {
  parentId?: IdType
  menuName?: string
  menuCode?: string
  type?: number
  status?: number
}

export interface MenuPageQueryParams {
  menuName?: string
  menuCode?: string
  type?: number
  status?: number
  pageNum?: number
  pageSize?: number
}

function buildQuery(params: Record<string, unknown>) {
  const query = new URLSearchParams()
  Object.entries(params).forEach(([key, value]) => {
    if (value !== undefined && value !== null && value !== '') {
      query.append(key, String(value))
    }
  })
  return query.toString()
}

export function listMenuTree(): Promise<ApiEnvelope<MenuTreeNode[]>> {
  return request<MenuTreeNode[]>('/api/menu/tree', { method: 'GET' })
}

export function listMenus(params: MenuListParams = {}): Promise<ApiEnvelope<Menu[]>> {
  const query = buildQuery(params)
  return request<Menu[]>(`/api/menu/list${query ? `?${query}` : ''}`, { method: 'GET' })
}

export function pageMenus(params: MenuPageQueryParams = {}): Promise<ApiEnvelope<PageResult<Menu>>> {
  const query = buildQuery(params)
  return request<PageResult<Menu>>(`/api/menu/page${query ? `?${query}` : ''}`, { method: 'GET' })
}

export function getMenu(id: IdType): Promise<ApiEnvelope<Menu>> {
  return request<Menu>(`/api/menu/${id}`, { method: 'GET' })
}

export function addMenu(payload: Partial<Menu>): Promise<ApiEnvelope<Menu>> {
  return request<Menu>('/api/menu/add', {
    method: 'POST',
    body: payload
  })
}

export function updateMenu(payload: Partial<Menu>): Promise<ApiEnvelope<void>> {
  return request<void>('/api/menu/update', {
    method: 'PUT',
    body: payload
  })
}

export function deleteMenu(id: IdType): Promise<ApiEnvelope<void>> {
  return request<void>(`/api/menu/${id}`, { method: 'DELETE' })
}
